op-node/rollup/derive/blob_data_source.go

```diff
@@ -0,0 +1,172 @@
+package derive
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+type blobOrCalldata struct {
+	// union type. exactly one of calldata or blob should be non-nil
+	blob     *eth.Blob
+	calldata *eth.Data
+}
+
+// BlobDataSource fetches blobs or calldata as appropriate and transforms them into usable rollup
+// data.
+type BlobDataSource struct {
+	data         []blobOrCalldata
+	ref          eth.L1BlockRef
+	batcherAddr  common.Address
+	dsCfg        DataSourceConfig
+	fetcher      L1TransactionFetcher
+	blobsFetcher L1BlobsFetcher
+	log          log.Logger
+}
+
+// NewBlobDataSource creates a new blob data source.
+func NewBlobDataSource(ctx context.Context, log log.Logger, dsCfg DataSourceConfig, fetcher L1TransactionFetcher, blobsFetcher L1BlobsFetcher, ref eth.L1BlockRef, batcherAddr common.Address) DataIter {
+	return &BlobDataSource{
+		ref:          ref,
+		dsCfg:        dsCfg,
+		fetcher:      fetcher,
+		log:          log.New("origin", ref),
+		batcherAddr:  batcherAddr,
+		blobsFetcher: blobsFetcher,
+	}
+}
+
+// Next returns the next piece of batcher data if any remains. It returns ResetError if it cannot
+// find the referenced block or a referenced blob, or TemporaryError for any other failure to fetch
+// a block or blob.
+func (ds *BlobDataSource) Next(ctx context.Context) (eth.Data, error) {
+	if ds.data == nil {
+		var err error
+		if ds.data, err = ds.open(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(ds.data) == 0 {
+		return nil, io.EOF
+	}
+
+	next := ds.data[0]
+	ds.data = ds.data[1:]
+	if next.calldata != nil {
+		return *next.calldata, nil
+	}
+
+	data, err := next.blob.ToData()
+	if err != nil {
+		ds.log.Error("ignoring blob due to parse failure", "err", err)
+		return ds.Next(ctx)
+	}
+	return data, nil
+}
+
+// open fetches and returns the blob or calldata (as appropriate) from all valid batcher
+// transactions in the referenced block. Returns an empty (non-nil) array if no batcher
+// transactions are found. Returns nil array whenever an error is returned.
```

Comment:
removed. I usually like to document where nil values might be returned to avoid panics but for slices I guess the expectation is nil might always be returned instead of the 0 value.

Reply:
Thanks! Yes Go has this thing where it treats nil vs empty slices the same.